larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,470 @@
+            float initialFigureOfMerit=this->GetFigureOfMerit(caloHitList3D);
+			
+			//Create a variable for the minimum figure of merit and initialize to initial FOM
+            float minimumFigureOfMerit(initialFigureOfMerit);
+
+            //Free the hits in this cluster, so that they are not owned by the original pfo, and are available for reclustering
+            PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RemoveFromPfo(*this, pShowerPfo, clusterList3D.front()));
+
+            // Initialize reclustering with these local lists
+            std::string currentClustersListName;
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Cluster>(*this, currentClustersListName));
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Cluster>(*this, "ShowerClusters3D"));
+
+			//I want to split the calo hit list into a new set of calo hit lists, taking the best outcome out of different algorithms
+            std::vector<CaloHitList*> newCaloHitListsVector, minimumFigureOfMeritCaloHitListsVector, initialCaloHitListsVector;
+            initialCaloHitListsVector.emplace_back(&caloHitList3D);
+            minimumFigureOfMeritCaloHitListsVector = initialCaloHitListsVector;
+
+            for (auto toolIter = m_algorithmToolVector.begin(); toolIter != m_algorithmToolVector.end(); ++toolIter)
+            {
+                newCaloHitListsVector = initialCaloHitListsVector;
+
+                try {
+                    (*toolIter)->Run(this, newCaloHitListsVector);
+                } catch (const StatusCodeException &){
+                    std::cout << "Exception caught! Cannot run reclustering tool!" << std::endl;
+                }
+
+                //Calculate FOM for this vector of new CaloHitLists
+				float newFigureOfMerit = this->GetFigureOfMerit(newCaloHitListsVector);
+
+                //Is this FOM smaller?
+                if(newFigureOfMerit <= minimumFigureOfMerit)
+                {
+                     minimumFigureOfMerit = newFigureOfMerit;
+                     minimumFigureOfMeritCaloHitListsVector = newCaloHitListsVector;
+                }
+                else //If not, clear the lists and vector!!
+                {
+                    for (CaloHitList* caloHitList : newCaloHitListsVector) delete caloHitList;
+                }
+                newCaloHitListsVector.clear();
+            }
+            
+            //If the new best calo hit lists outcome is equivalent to original, move pfo to unchanged pfo list. Else, create new vector of 3D clusters
+            if(minimumFigureOfMeritCaloHitListsVector==newCaloHitListsVector)
+            {
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::AddToPfo(*this, pShowerPfo, clusterList3D.front()));
+                unchangedPfoList.push_back(pShowerPfo);
+            }
+            else
+            {
```

Comment:
I'd be tempted to drop the indented else block and make the `if` part have a continue at the end.

That way, the code isn't quite as indented which is a pain when there is the very long `PandoraContentApi` and `PANDORA_RETURN_RESULT_IF` lines, which are basically unavoidable.

Reply:
Good idea, I've changed this.